### PR TITLE
[8.x] Consistent boolean behaviour

### PR DESCRIPTION
**Pull request by rihardsgrislis**
Currently `boolean` validation rule differs from how request's boolean helper method works. This PR aligns the behaviour of both. More at: #33692.

PR for docs: https://github.com/laravel/docs/pull/6210

### CODE REVIEW

**Comment by rihardssceredins** [comment]
@GrahamCampbell hey, got the tests passing however.. It appears there is a unresolved (and mistakenly closed) bug with PHP's filter_var: https://bugs.php.net/bug.php?id=68761.

`filter_var(null, FILTER_VALIDATE_BOOLEAN, FILTER_NULL_ON_FAILURE)` currently returns `false` when per documentation it should return `null` because `null` itself isn't valid value. I've added a patch for that specific case and added assertion to check that.

What do you think?

**Comment by GrahamCampbell** [comment]
Why not change the other code to match this bug-free validator code, rather than to change the validator code?

**Comment by rihardssceredins** [comment]
By "other code" you mean the previous implementation with  `$acceptable = [true, false, 0, 1, '0', '1']; `?

**Comment by rihardssceredins** [comment]
If so then I was thinking about it but idk.. wouldn't we be back at square one with two implementations (one for validation and the other for casting)?

**Comment by taylorotwell** [comment]
Can you summarize the breaking changes here? New values that would be considered valid? Old valid values that would now be considered invalid?

**Comment by laurencei** [comment]
Hi. I did the original PR for the `boolean` validation rule back in 2014 (my first ever PR to Laravel!): https://github.com/laravel/framework/pull/4599

There is a very explicit reason for using `true, false, 1, 0, "1", and "0"` as the rules and not `FILTER_VAR`

The problem is you cannot accept `"false"` as `false` - because `"false" === true`. The string `"false"` is no more a boolean value than `"nah"`, `"nope"` or `"disagree"`. They are all negative english words, but mean nothing in the land of PHP.

Consider this code:

```
$user = User::find(1);
$user->accepted = "false";
$user->save();
```

On my dev environment - this throws a SQL error:

```
SQLSTATE[HY000]: General error: 1366 Incorrect integer value: 'false' for column 'accepted' at row 1 (SQL: update `users` set `accepted` = false, `users`.`updated_at` = 2020-07-29 23:56:17 where `id` = 1)
```

But the other risk is SQL will typecast "false", and then you end up with this:

```
$new = User::find(1);
dd($new->accepted);  // Shows 1, which is "true", which is incorrect.
```

There main issue here is FILTER_VAR will typecast and return `"false"` as `false` as part of the PHP function - but because the validation engine doesnt then overrwrite and typecast the variable in the request object, this doesnt help us.

The better option here is middleware that typecasts any "false" to `false`, then let it pass through the system as normal. That way you are actually dealing with a boolean value, not a string representation of a boolean.

This question comes up every now and again. Maybe shipping some optional middleware (disabled by default) in the Laravel package could give people the option to enable it when they believe then need it.

**Comment by taylorotwell** [comment]
I think we will not switch to filter_var based on @laurencei's feedback.
